File: jediterm-core-mpp/src/jvmMain/java/com/jediterm/terminal/model/TerminalSelection.java
package com.jediterm.terminal.model;

import com.jediterm.core.compatibility.Point;
import kotlin.Pair;
import org.jetbrains.annotations.Nullable;

/**
 * @author traff
 */
public class TerminalSelection {
  private final Point myStart;

  private Point myEnd;

  public TerminalSelection(Point start) {
    myStart = start;
  }

  public TerminalSelection(Point start, Point end) {
    myStart = start;
    myEnd = end;
  }

  public Point getStart() {
    return myStart;
  }

  public Point getEnd() {
    return myEnd;
  }

  public void updateEnd(Point end) {
    myEnd = end;
  }

  public Pair<Point, Point> pointsForRun(int width) {
    Pair<Point, Point> p = SelectionUtil.sortPoints(new Point(myStart), new Point(myEnd));
    p.getSecond().x = Math.min(p.getSecond().x + 1, width);
    return p;
  }

  public boolean contains(Point toTest) {
    return intersects(toTest.x, toTest.y, 1);
  }

  public void shiftY(int dy) {
    myStart.y += dy;
    myEnd.y += dy;
  }

  public boolean intersects(int x, int row, int length) {
    return null != intersect(x, row, length);
  }

  @Nullable
  public Pair<Integer, Integer> intersect(int x, int row, int length) {
    int newX = x;
    int newLength;

    Pair<Point, Point> p = SelectionUtil.sortPoints(new Point(myStart), new Point(myEnd));

    if (p.getFirst().y == row) {
      newX = Math.max(x, p.getFirst().x);
    }

    if (p.getSecond().y == row) {
      newLength = Math.min(p.getSecond().x, x + length - 1) - newX + 1;
    } else {
      newLength = length - newX + x;
    }

    if (newLength<=0 || row < p.getFirst().y || row > p.getSecond().y) {
      return null;
    } else
      return new Pair<>(newX, newLength);
  }

  @Override
  public String toString() {
    return "[x=" + myStart.x + ",y=" + myStart.y + "]" + " -> [x=" + myEnd.x + ",y=" + myEnd.y + "]";
  }

}

File: jediterm-core-mpp/src/jvmMain/kotlin/com/jediterm/terminal/model/TerminalSelection.kt
package com.jediterm.terminal.model

import com.jediterm.core.compatibility.Point
import kotlin.math.max
import kotlin.math.min

/**
 * @author traff
 */
class TerminalSelection {
    val start: Point

    private var myEnd: Point? = null

    constructor(start: Point) {
        this.start = start
    }

    constructor(start: Point, end: Point) {
        this.start = start
        myEnd = end
    }

    val end: Point?
        get() = myEnd

    fun updateEnd(end: Point) {
        myEnd = end
    }

    fun pointsForRun(width: Int): Pair<Point?, Point?> {
        val endPoint = myEnd ?: return Pair(Point(start), null)
        val p = SelectionUtil.sortPoints(
            Point(start),
            Point(endPoint)
        )
        p.second?.let { it.x = min(it.x + 1, width) }
        return p
    }

    fun contains(toTest: Point): Boolean {
        return intersects(toTest.x, toTest.y, 1)
    }

    fun shiftY(dy: Int) {
        start.y += dy
        myEnd?.let { it.y += dy }
    }

    fun intersects(x: Int, row: Int, length: Int): Boolean {
        return null != intersect(x, row, length)
    }

    fun intersect(x: Int, row: Int, length: Int): Pair<Int?, Int?>? {
        val endPoint = myEnd ?: return null
        var newX = x
        val newLength: Int

        val p = SelectionUtil.sortPoints(
            Point(start),
            Point(endPoint)
        )

        val first = p.first ?: return null
        val second = p.second ?: return null

        if (first.y == row) {
            newX = max(x, first.x)
        }

        if (second.y == row) {
            newLength = min(second.x, x + length - 1) - newX + 1
        } else {
            newLength = length - newX + x
        }

        if (newLength <= 0 || row < first.y || row > second.y) {
            return null
        } else return Pair<Int?, Int?>(newX, newLength)
    }

    override fun toString(): String {
        return myEnd?.let {
            "[x=" + start.x + ",y=" + start.y + "]" + " -> [x=" + it.x + ",y=" + it.y + "]"
        } ?: "[x=" + start.x + ",y=" + start.y + "] -> [no end]"
    }
}
